Козырев Д.В., "Кватернион", 3530901/10001

kozyrev-dv opened the pull request.

Commit:
Ready to check work

h31 commented on src/main/java/Quaternion.java:
```diff
@@ -0,0 +1,389 @@
+import java.lang.invoke.VarHandle;
+import java.security.InvalidParameterException;
+
+/* Артём Олегович, здравствуйте!
```
Здравствуйте, Даниил. Работа правда сделана очень добротно.

Давайте сначала обсудим организационный момент. Если хотите оставить для меня вопрос по коду, используйте для этого Гитхаб. Комментарии в коде для этого лучше не использовать - я отвечу на вопрос, потом вопрос нужно будет убрать, потребуется изменить код - это неаккуратно и не очень удобно.

Хороший вопрос. Универсального ответа тут нет, в целом по возможности лучше делать иммутабельные [value object](https://en.wikipedia.org/wiki/Value_object). В новых версиях Java для этого даже добавили конструкцию `record`. Мутабельные объекты стоит использовать там, где без них никак или когда это позволяет заметно повысить эффективность решения. Почитать по теме: [раз](https://javarush.ru/quests/lectures/questmultithreading.level02.lecture01), [два](https://stackoverflow.com/questions/1284727/mutable-or-immutable-class).

`static` стоит использовать, если вам точно в этом методе не нужны поля класса или так код получается чище, в остальных случаях лучше писать обычные методы. [Материал](https://stackoverflow.com/questions/2671496/when-to-use-static-methods) по теме.

h31 replied:
В любом случае, выбрать стоит один вариант и убрать остальные. Сейчас у вас есть dead code в проекте, от него стоит избавиться.

h31 replied:
P.S. `моветон`

h31 commented on src/main/java/Vector3.java:
```diff
@@ -0,0 +1,291 @@
+
+public class Vector3 {
+
+    private static final double equivalenceDelta = 1e-13;
```
Такая же константа есть в другом файле

h31 commented on src/main/java/Quaternion.java:
```diff
@@ -0,0 +1,389 @@
+* если это мовитон?
+* */
+
+public class Quaternion {
```
Представим, что я хочу написать игру, и мне понадобился ваш кватернион. Написал такой код:
```java
package ru.spbstu.insys.game;

public class View {
    Quaternion quaternion = new Quaternion();
}
```
Не работает. Почему?

kozyrev-dv replied:
К сожалению, единственное что мне приходит в голову, это то, что Вы не включили в код пакет, в котором реализован кватернион. Я буквально еще раз перепроверил, и IDE никаких ошибок компиляции и выполнения не дает

h31 replied:
<img width="992" alt="image" src="https://user-images.githubusercontent.com/3375923/155706265-54a6ec26-91a4-436e-ae95-587bd7daabc0.png">

h31 commented on src/main/java/Vector3.java:
```diff
@@ -0,0 +1,291 @@
+
+        final Vector3 other = (Vector3) obj;
+
+        validateVector(other);
```
Насколько другой разработчик может ожидать, что `equals` выбросит исключение?

h31 commented on src/main/java/Vector3.java:
```diff
@@ -0,0 +1,291 @@
+            throw new InvalidParameterException("Infinite value is not allowed");
+        }
+    }
+    private static void validateVector(@NotNull Vector3 vector) {
```
Не забудьте сделать `Reformat code`

h31 commented on src/main/java/Vector3.java:
```diff
@@ -0,0 +1,291 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.security.InvalidParameterException;
```
Уверены, что ваша ошибка относится к безопасности (`security`)?

h31 commented on src/main/java/Vector3.java:
```diff
@@ -0,0 +1,291 @@
+
+        validateVector(other);
+
+        // Не упрощаю по предложению IDE, потому что кажется, что так понятнее
```
В этом коде явно можно вынести общую часть. Возможно, после этого можно будет воспользоваться рекомендацией IDE без потери читаемости.

h31 commented on src/main/java/Vector3.java:
```diff
@@ -0,0 +1,291 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Если переопределяете `equals`, то нужно переопределять и `hashCode`, см. контракт к `hashCode`

h31 commented on src/main/java/Quaternion.java:
```diff
@@ -0,0 +1,389 @@
+        }
+    }
+
+    public static double getAngleFromW(double w) {
```
Больше похоже на `private` метод

h31 commented on src/main/java/Quaternion.java:
```diff
@@ -0,0 +1,389 @@
+        sb.append(w);
+        if (x < 0) {
+            sb.append(" - ");
+        } else sb.append(" + ");
```
Лучше ставить фигурные скобки симметрично, так код будет проще читаться

h31 commented on src/main/java/Quaternion.java:
```diff
@@ -0,0 +1,389 @@
+        sb.append("Quaternion: ");
+        sb.append(w);
+        if (x < 0) {
+            sb.append(" - ");
```
Вам здесь явно пригодился бы `String.format`

h31 commented on src/test/java/QuaternionTest.java:
```diff
@@ -0,0 +1,211 @@
+        // сравню папарно каждый элемент 2-х кватернионов в отдельных тестах.
+        Quaternion q1 = new Quaternion(3, 2.1, -2, -2.5).reciprocal();
+        Quaternion q2 = new Quaternion(0.126796, -0.088757, 0.084531, 0.105664);
+        assertEquals(q2.getW(), q1.getW(), 5e-7);
```
У вас есть константа в коде, которую можно использовать вместо `5e-7`

h31 commented on src/test/java/QuaternionTest.java:
```diff
@@ -0,0 +1,211 @@
+        // 1.6 / 3.0724582991474434 = 0.5207556439232954
+        // -1.2 / 3.0724582991474434 = -0.3905667329424716
+        // -2 / 3.0724582991474434 = -0.6509445549041193
+        assertEquals(new Quaternion(0.3905667329424716, 0.5207556439232954, -0.3905667329424716, -0.6509445549041193),
```
Есть предложение, как избавиться от этих длиннющих чисел. Вам пришла в голову идея взять калькулятор как эталонную реализацию - это очень хорошая мысль. Давайте пойдем чуть дальше, для Java есть существующие реализации кватернионов, например, в Apache Commons Math. Стоит проверить в рамках автоматических тестов, что ваша реализация даёт такие же результаты, что и эталонная.

h31 commented on src/test/java/QuaternionTest.java:
```diff
@@ -0,0 +1,211 @@
+        // q₁ = 3 + 2.1i - 2j - 2.5k
+        // q₁⁻¹ = 0.126796 - 0.088757i + 0.084531j + 0.105664k.
+        // Т.к. сравнивать кватернионы с выборочной точностью мне представляется невозможным,
+        // сравню папарно каждый элемент 2-х кватернионов в отдельных тестах.
```
`попарно`

h31 commented on src/main/java/Quaternion.java:
```diff
@@ -0,0 +1,389 @@
+    }
+
+    public void setW(double w) {
+        validateInput(w);
```
Валидировать данные - это, безусловно, правильно и полезно. Вопрос вызывает другой момент - нужно ли это делать каждый раз, когда мы создаем новый объект. С одной стороны, ошибки лучше обнаруживать как можно раньше (подход [fail fast](https://en.wikipedia.org/wiki/Fail-fast)). С другой, это накладно в плане производительности - выполнять одну и ту же проверку много раз над одинаковыми данными. Кроме того, в таком случае практически любой метод кватерниона может выбросить исключение (мы этого не замечаем, потому что оно unchecked), и не всегда в вызывающем коде мы можем понять, как его обрабатывать. Подумайте, как лучше сделать и какой вариант более предпочтительный в данной задаче.

kozyrev-dv replied:
Я принял решение вообще не делать проверок. Вектор бесконечной длины может быть, а все операции с NaN, вернут NaN, поэтому будет понятно, что что-то идет не так. Для мануальной проверки "Все ли хорошо", просто перепишу метод validateInput. Думаю, это хороший вариант, т.к. все внутренние операции классов сделаны сейчас безопасно. Так что просто чтобы проследить за состоянием объекта можно будет проверять валидность данных.

Commit:
1. Engine.java -> Useful constants are pulled into.
2. pom.xml -> new maven dependency added.
3. Quaternion.java:
---3.1. Moved into engine.math package;
---3.2. Code cleaned; Quaternion is now value object;
---3.3. No error checks now, value validation is manual now (use: tryValidate());
---3.4. equals() doesn't throw errors now;
---3.5. toString() refactored.
4. Vector3.java:
---4.1. Moved into engine.math package;
---4.2. Code cleaned; Vector3 is now value object;
---4.3. No error checks now, value validation is manual now (use: tryValidate());
---4.4. equals() doesn't throw errors now;
---4.5. toString() refactored;
---4.6. rotated() functionality changed.
5.QuaternionTest.java -> All tests rewritten

kozyrev-dv commented:
Здравствуйте, Артем Олегович!

Переделал задание по всеми Вашим комментариями, жду Вашего мнения :)
Кажется, что переделано много, но на самом деле классы Vector3 и Quaternion, почти не изменены. Подробно о изменениях в работе написано в комментарии к коммиту.

Commit:
1. Quaternion.java:
---1.1. Explicit code deleted;
---1.2. Exception type changed to IllegalArgumentException;
---1.3. Explicit "this" keywords deleted;
---1.4. Return statement in equals() cleared;
---1.5. hashCode() function overridden.
2. Engine.java:
---2.1. Renamed to Utils.java;
---2.2. Double comparison added.
3. Vector3.java:
---3.1. Explicit code deleted;
---3.2. Exception type changed to IllegalArgumentException;
---3.3. Explicit "this" keywords deleted;
---3.4. Return statement in equals() cleared;
---3.5. hashCode() function overridden.
4. QuaternionTest.java -> tests added; full coverage completed.

h31 commented on src/test/java/QuaternionTest.java:
```diff
@@ -0,0 +1,297 @@
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+
+import java.security.InvalidParameterException;
```
Лишний импорт

h31 commented on src/test/java/QuaternionTest.java:
```diff
@@ -0,0 +1,297 @@
+        assertTrue(q1.hashCode() == q2.hashCode());
+
+        assertNotEquals(new engine.math.Quaternion(0, 1, 3, 5), null);
+        assertNotEquals(new engine.math.Quaternion(0, 3, 5, 6), new Vector3(0, 3, 6));
```
Обратите внимание на результат сканирования

h31 commented on src/test/java/QuaternionTest.java:
```diff
@@ -0,0 +1,297 @@
+
+    @Test
+    void QuaternionEquals() {
+
```
Лишняя пустая строчка

h31 reviewed (approved):
Сейчас у вас падают тесты, их нужно исправить. Плюс, обратите внимание на предупреждения Qodana. Когда исправите, приму работу, оценка 5.

Commit:
QuaternionTest.java -> QuatenionToString() test Locale check added

Commit:
QuaternionTest.java -> QuatenionToString() test Locale check added

Commit:
QuaternionTest.java -> QuatenionEquals() and Vector3Equals() tests changed
